File: src/main/cpp/min_distance.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <sstream>
#include <stack>

using namespace std;

string find_min_distance();


int main() {
    cout << find_min_distance() << endl;
    return 0;
}

std::string find_min_distance() {
    int vertexes, edges;
    cin >> vertexes >> edges;
    unordered_map<int, vector<int> > graph;
    for (int i = 0; i < edges; i++) {
        int from, to;
        cin >> from >> to;
        vector<int> adjacent;
        if (graph.find(from) != graph.end()) {
            graph.at(from).push_back(to);
        } else {
            adjacent.push_back(to);
            graph.insert(pair<int, vector<int> >(from, adjacent));
        }

        vector<int> inversed_adjacent;
        if (graph.find(to) != graph.end()) {
            graph.at(to).push_back(from);
        } else {
            inversed_adjacent.push_back(from);
            graph.insert(pair<int, vector<int> >(to, inversed_adjacent));
        }
    }
    int start_node, to_find;
    cin >> start_node >> to_find;

    bool visited[vertexes + 1];
    for (int i = 0; i < vertexes + 1; ++i) {
        visited[i] = false;
    }

    deque<pair<int, int> > deque;
    deque.push_back(pair < int, int > (start_node, 0));
    while (!deque.empty()) {
        auto p = deque.front();
        deque.pop_front();
        if (p.first == to_find) {
            return to_string(p.second);
        }
        if (visited[p.first]) {
            continue;
        }
        visited[p.first] = true;
        if (graph.find(p.first) != graph.end()) {
            for (int vertex: graph.at(p.first)) {
                deque.push_back(pair < int, int > (vertex, p.second + 1));
            }
        }

    }
    return to_string(-1);
}

File: src/main/java/MinDistance.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.*;

public class MinDistance {

    public static void main(String[] args) throws IOException {
        System.out.println(wrap(System.in));
    }

    public static String wrap(InputStream is) throws IOException {
        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(is))) {
            final String[] firstLine = reader.readLine().split(" ");
            final int vertexes = Integer.parseInt(firstLine[0]);
            final int edges = Integer.parseInt(firstLine[1]);
            final boolean[][] graph = new boolean[vertexes + 1][vertexes + 1];
            for (int i = 0; i < edges; i++) {
                final Integer[] edge = Arrays.stream(reader.readLine().split(" "))
                        .map(Integer::parseInt)
                        .toArray(Integer[]::new);
                graph[edge[0]][edge[1]] = true;
                graph[edge[1]][edge[0]] = true;
            }
            final String[] lastLine = reader.readLine().split(" ");
            final int startNode = Integer.parseInt(lastLine[0]);
            final int toFind = Integer.parseInt(lastLine[1]);

            final Set<Integer> visited = new HashSet<>(vertexes);
            final Deque<Pair> queue = new LinkedList<>();
            queue.addLast(new Pair(startNode, 0));
            while (!queue.isEmpty()) {
                final Pair curNode = queue.pop();

                if (curNode.node.equals(toFind)) {
                    return curNode.dist.toString();
                }
                if (visited.contains(curNode.node)) {
                    continue;
                }
                visited.add(curNode.node);

                for (int i = 0; i <= vertexes; i++) {
                    if (graph[curNode.node][i]) {
                        queue.addLast(new Pair(i, curNode.dist + 1));
                    }
                }
            }
            return String.valueOf(-1);
        }
    }

    private static class Pair {
        public final Integer node;
        public final Integer dist;

        private Pair(Integer node, Integer dist) {
            this.node = node;
            this.dist = dist;
        }
    }
}
